src/audio/mixer.c

```diff
@@ -562,22 +580,25 @@ void mixer_exec(int32_t *out, int num_samples) {
 		}
 	}
 
-	// Copy the volumes into DMEM. TODO: check if should change this loop into
-	// a DMA copy, or fold it into the above loop.
 	uint32_t *lvol32 = (uint32_t*)lvol;
 	uint32_t *rvol32 = (uint32_t*)rvol;
 	for (int ch=0;ch<MIXER_MAX_CHANNELS/2;ch++)  {
-		SP_DMEM[4+0*16+ch] = lvol32[ch];
-		SP_DMEM[4+1*16+ch] = rvol32[ch];
+		settings->lvol[ch] = lvol32[ch];
+		settings->rvol[ch] = rvol32[ch];
 	}
 
-	SP_DMEM[0] = MIXER_FX16(Mixer.vol);
-	SP_DMEM[1] = (num_samples << 16) | Mixer.num_channels;
-	SP_DMEM[2] = (uint32_t)out;
-	SP_DMEM[3] = (uint32_t)Mixer.ucode_state;
-
 	uint32_t t0 = TICKS_READ();
-	rsp_run();
+
+	rspq_highpri_begin();
+	rspq_write(__mixer_overlay_id, 0,
+		(((uint32_t)MIXER_FX16(Mixer.vol)) & 0xFFFF),
```

Comment:
How do we make sure MSB is all zeros here? (now knowing mixer ucode yet)

Reply:
Mixer.vol is supposed to be in the 0..1 range. If it can't go beyond that it's a bug, but I wouldn't clamp it here -- I would do that where it can wrongly assume that out-of-bound value. This is the inner execution loop where the global state is assumed to be consistent and valid.